### PR TITLE
Add columns during edit

### PR DESCRIPTION
**Pull request by squirrelo**
Since this is needed for EBI submission, this should get reviewed and merged quickly.

This allows addition of columns during editing of a sample template. It also cleans up the add and remove category functions, as they were missing the addition of the column names to the study_sample_columns table.

### CODE REVIEW

**Comment by josenavas** [comment]
@squirrelo besides the problems connecting to the redis server, it looks like there is a legitimate test failure!

**Comment by squirrelo** [comment]
Seems like only redis failures now.

**Comment by squirrelo** [comment]
ready for review

File: qiita_db/metadata_template.py [hunk 1]
```diff
@@ -1672,16 +1686,29 @@ def add_category(self, category, samples_and_values, dtype, default):
...
+        queue = 'add_col_%d' % self._id
+        conn_handler.create_queue(queue)
+
+        # add the column to the sample template
+        sql = """ALTER TABLE qiita.{0}
+            ADD COLUMN {1} {2}""".format(table_name, category, dtype)
+        if default is not None:
+            # mogrify the sql and add to end
+            with conn_handler.get_postgres_cursor() as cur:
+                sql = sql + cur.mogrify(" NOT NULL DEFAULT %s", [default])
```
Comment (josenavas) [inline qiita_db/metadata_template.py]: Check the comment of the first answer on this [thread](http://stackoverflow.com/q/6775497/3746629). In short, mogrify should be used only for debugging, not for generating code. Anyways, I also find this solution more convoluted, so if you can just add the if and generate either of the sql in the specific case of the if would be more cleaner/maintainable.

File: qiita_db/metadata_template.py [hunk 2]
```diff
@@ -1561,9 +1561,16 @@ def update(self, md_template):
                                % ', '.join(samples_diff))
         columns_diff = set(new_map.columns) - set(current_map.columns)
         if columns_diff:
-            raise QiitaDBError('The new sample template differs from what is '
-                               'stored in database by these columns names: %s'
-                               % ', '.join(columns_diff))
+            queue = 'add_cols_%d' % self._id
```
Comment (josenavas) [inline qiita_db/metadata_template.py]: A general comment: using the approach of multiple queues on a single function, we still can face the problem in which half of the function is successfully applied while the other half is not, leaving the system in a inconsistent state. Can you use a single queue in the entire function?

Thanks!

File: qiita_db/metadata_template.py [hunk 3]
```diff
@@ -1561,9 +1561,16 @@ def update(self, md_template):
...
+            datatypes = dict(izip(new_map.columns, _get_datatypes(new_map)))
+            conn_handler.create_queue(queue)
+
+            for col in columns_diff:
+                self.add_category(col, new_map[col].to_dict(), datatypes[col])
```
Comment (josenavas) [inline qiita_db/metadata_template.py]: This is going to leverage a lot of connection openings/closing to the database, with the added danger of leveraging the db in an inconsistent state. You can either (a) duplicate the code of that function here so everything is executed in a single queue; or (b) modify the function signature so a conn_handler and a queue name can be passed, so it just gets added to the queue; or (c) do a private function to add things to a queue which is shared by both functions.

My vote is (c) as that pattern has been already done in other portions of the code (the data objects) and it is the one that maintains a consistency on the code and in the functions (i.e. each function only has a possible output w/o side effects depending on the input parameters)

Comment (squirrelo) [inline qiita_db/metadata_template.py]: This is how the old code did it (look at line 1587). If we want a full re-factor that is out of the scope of this pull request, we just need the functionality quickly for EBI submissions. Issue .#1020 created for a full refactor.

File: qiita_db/metadata_template.py (hunk 2 above)
Comment (squirrelo) [inline qiita_db/metadata_template.py]: This is consistent with the existing codebase, and a refactor is out of scope. Issue #1020 created for the refactor.

File: qiita_db/metadata_template.py (hunk 3 above)
Comment (josenavas) [inline qiita_db/metadata_template.py]: Thx

**Comment by josenavas** [comment]
Closing in favor of #1021

**Pull request closed without merging**